### PR TITLE
Fix `Hashable` not working with `model_validate_json`

### LINKED ISSUE #9567: `model_validate_json` cannot validate `Hashable`
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

image

### Example Code

```Python
from typing import Hashable

from pydantic import BaseModel

class A(BaseModel):
    x: Hashable

a = A.model_validate_json('{"x": "1"}')
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.7.0
        pydantic-core version: 2.18.1
          pydantic-core build: profile=release pgo=true
                 install path: D:\ProgramData\miniconda3\Lib\site-packages\pydantic
               python version: 3.11.4 | packaged by Anaconda, Inc. | (main, Jul  5 2023, 13:47:18) [MSC v.1916 64 bit (AMD64)]
                     platform: Windows-10-10.0.22631-SP0
             related packages: email-validator-2.1.0.post1 fastapi-0.100.0 mypy-1.4.1 pydantic-extra-types-2.1.0 pydantic-settings-2.1.0 pyright-1.1.331 typing_extensions-4.8.0
                       commit: unknown
```

### PR DESCRIPTION
**Pull request by kc0506**
## Change Summary

Wrap a `json-or-python` schema for `Hashable` type. This allows `validate_json` to work consistently with validating Python objects.

This breaks some existing tests, though. Currently my concern is about `model_dump_json`: when a hashable field has type other than `Union[str, int, float, bool, none]` (e.g. `tuple`), the model can still be serialized (with warning), but the dumped json cannot be validated by the model, which seems inconsistent. I am thinking just raising error when serializing such fields; but I couldn't find a way to do it.

### Example
```py
class A(BaseModel):
    x: Hashable

a = A.model_validate_json('{"x": "1"}')
```

## Related issue number

fix #9567

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by sydney-runkle** [comment]
Hey @kc0506, thanks for the PR! I'll review first thing tomorrow 🚀 !

File: pydantic/_internal/_generate_schema.py [hunk 1]
```diff
@@ -1741,8 +1741,21 @@ def _pattern_schema(self, pattern_type: Any) -> core_schema.CoreSchema:
...
 
     def _hashable_schema(self) -> core_schema.CoreSchema:
+        # json cannot be validated by isinstance, so we manually construct a schema
+        json_schema = core_schema.union_schema(
+            [
+                core_schema.str_schema(),
+                core_schema.int_schema(),
+                core_schema.float_schema(),
+                core_schema.bool_schema(),
+                core_schema.none_schema(),
```
Comment (sydney-runkle) [inline pydantic/_internal/_generate_schema.py:1751]: What about `tuple`, `frozenset`, or other types like that? There might be a few more we want to include. I suppose we can accept slow validation here given that the type is so broad?

File: pydantic/_internal/_generate_schema.py [hunk 2]
```diff
@@ -1741,8 +1741,21 @@ def _pattern_schema(self, pattern_type: Any) -> core_schema.CoreSchema:
...
+            ]
+        )
         return core_schema.custom_error_schema(
-            core_schema.is_instance_schema(collections.abc.Hashable),
+            core_schema.json_or_python_schema(
+                json_schema=json_schema,
+                python_schema=core_schema.is_instance_schema(collections.abc.Hashable),
+            ),
```
Comment (sydney-runkle) [inline pydantic/_internal/_generate_schema.py:1758]: We could use a custom `serialization` schema, see https://github.com/pydantic/pydantic-core/blob/57e69918670016c4cf00a844aabcf76bb0cd458f/python/pydantic_core/core_schema.py#L427

Comment (sydney-runkle) [reply pydantic/_internal/_generate_schema.py:1758]: Here's an example: https://github.com/pydantic/pydantic/blob/1e413857f48a807d3c81281b52c571f07c7077ee/pydantic/_internal/_generate_schema.py#L1703-L1708

**Review by sydney-runkle — COMMENTED** [review]
Looking good thus far!

**Comment by kc0506** [comment]
I've updated both the json and python schemas with some notable changes:

For json schema, Non-primitive types are now considered hashable:
- lists are validated as tuple.
- objects are validated as `frozendict`.
Note that python doesn't seem to have built-in immutable and hashable mapping type, so I have naively implemented one. It is temporarily defined inside `_hashable_schema` function because I am not sure where to put it.

The enhanced json schema seems to be more flexible than python schema, which feels weird to me. So I add more choices to the python schema if input is not an instance of `Hashable`:
1. Convert `Sequence` types into `tuple`.
2. Attempt to transform input into `frozendict` (works with `BaseModel` since `dict(BaseModel())` is valid).
3. Try converting input into `frozenset`.

These changes increase schema flexibility but might be considered breaking since previously unhashable types (e.g., `list`) can now be validated as `Hashable`. I'm seeking feedback on whether this direction is appropriate or if adjustments are needed.

**Comment by kc0506** [comment]
Seems weird that tests failed. The failed case does not reproduce on my machine🤔

**Comment by sydney-runkle** [comment]
> For json schema, Non-primitive types are now considered hashable

Hmm, I see where you're coming from here. I might turn to the team to get feedback from a few folks regarding this. I suppose we might want to use a lax and strict approach, as I've mentioned below.

> Note that python doesn't seem to have built-in immutable and hashable mapping type, so I have naively implemented one. It is temporarily defined inside _hashable_schema function because I am not sure where to put it.

Interesting, I'll take a look.

> The enhanced json schema seems to be more flexible than python schema, which feels weird to me. 

This is a common pattern throughout our codebase - json schemas are more lax / permit more data coercion to allow for successful round trips. It is a bit of an odd pattern, though.

> So I add more choices to the python schema if input is not an instance of Hashable

Hmm. I think in general, we want to be pretty strict (not do all of this under the hood coercion). If folks want that behavior, they can use custom validators. Though, we could use a strict or lax validator to allow for supporting the best of both worlds.

> I'm seeking feedback on whether this direction is appropriate or if adjustments are needed.

I'll give some line by line feedback 👍.

**Comment by sydney-runkle** [comment]
Happy to do the line by line reviews tomorrow morning :)

**Comment by sydney-runkle** [comment]
Hmph, still struggling with this. `Hashable` is so broad, that it's hard to make the right choice regarding lax / strict behavior for all of the various types that it might support. We have some trouble with `Sequence` in this regard.

Will chat with our team early next week about next steps here.

**Comment by sydney-runkle** [comment]
> Hmph, still struggling with this. Hashable is so broad, that it's hard to make the right choice regarding lax / strict behavior for all of the various types that it might support. We have some trouble with Sequence in this regard.

Following up on this - what if we make the `Hashable` schema more simple - via a json or python schema:
* For python, we just try `hash` on the input data, and if we don't get a type error, then validation succeeds
* For json, we basically make this an after validator on an `any` type, and do the same thing

This removes some of the decision making complexity, but admittedly rejects coercion that could be done with lax / strict. I'm keen to avoid that though, bc it's sort of a scary hole to go down in terms of recursive structures - like what happens if we have a list of dicts, just converting that to a tuple of dicts isn't sufficient, etc.

Also, if folks truly want to customize it beyond that simple pattern, they can create a custom type or validator 🤷‍♀️

**Pull request closed without merging**